haskell-auto-import.el

```diff
@@ -0,0 +1,286 @@
+           '(lambda (acc str)
+              (let ((cur (car acc))
+                    (rest (cdr acc))
+                    (strlist (string-to-list str)))
+                (if cur
+                    (if (member ?\) strlist)
+                        (cons () (cons (cons str cur) rest))
+                      (cons (cons str cur) rest))
+                  (if (and
+                       (member ?\( strlist)
+                       ;; If the close paren is present too, we have a whole
+                       ;; token
+                       (not (member ?\) strlist)))
+                      (cons (list str) rest)
+                    (cons () (cons (list str) rest))))))
+           exploded
+           :initial-value (list ()))))
+    (mapcar
+     '(lambda (xs)
+        (if (equal (string-to-char (car xs)) ?\()
+            xs
+          (reduce 'append
+                  (mapcar
+                   '(lambda (s)
+                      (split-string s "[()]" t "\\s-"))
+                   (reverse xs)))))
+     (reverse (cdr parts)))))
+
+(defun haskell-auto-import-module-single-import-info (module-name &optional start)
+  "Returns information about the first import of the given module after `start'
+(which is set to 0 if nil). This information is nil if the module is not
+imported. Otherwise it is a list consisting of `(list start-point end-point
+is-qualified qualified-name imported-items)' where:
+- `start-point' is the point at the start of the import line
+- `end-point' is the point at the end of the import line
+- `is-qualified' is `t' if the import is qualified, nil otherwise
+- `qualified-name' is nil if no `as' clause was specified, or the qualified name
+  otherwise.
+- `imported-items' is a list of items imported from the module, in the format
+  provided by `haskell-auto-import-parse-import-parens' if it exists, nil otherwise."
+  (let ((contents (buffer-string))
+        (regexp
+         (concat
+          ;; Match the import keyword followed by whitespace
+          "^import\\s-+"
+          ;; Optionally match the qualified keyword as group 1
+          "\\(qualified\\s-+\\)?"
+          ;; Match exactly the module name
+          (regexp-quote module-name)
+          ;; Optionally match an `as' qualifier as group 2
+          "\\(?:\\s-+as\\s-+\\(\\sw+\\)\\)?"
+          ;; Whitespace may or may not precede parentheses at the end
+          "\\s-*"
+          ;; Optionally match the contents of some following parentheses as
+          ;; group 3
+          "\\(?:(\\(\\(?:[a-zA-Z0-9', \t\f\n\r]*\\|(.*?)\\)*\\))\\)?"
+          )))
+    (if (string-match regexp contents start)
+        (let ((is-qualified (match-string 1 contents))
+              (qualified-name (match-string 2 contents))
+              (imported-items-str (match-string 3 contents))
+              (start-point (match-beginning 0))
+              (end-point (match-end 0)))
+          (list start-point end-point
+                (if is-qualified t nil)
```

Comment:
In elisp, any non-nil value is truthy, so you can replace this entire `if` expression with `is-qualified`.